Find frequency bands

aburrell opened the pull request:
The field-of-view determination code requires that frequency bands be processed separately to avoid frequency-dependent shifts in the location of different elevation angle values.  There are other reasons why people may want to explore the frequencies of fitACF data files, so I added this as its own executable. 

This partially addresses #135.  The function `get_fit_tfreq_bands` will need to be called by the future FoV routine.

To test:
```
find_freqband -bandwidth 500 20160130.dce.fitacf3 > dce.freq_band_500
```

This will produce an ASCII file named `dce.freq_band_500` that has two frequency bands of 500 kHz and 377 observed transmission frequencies contained in these two bands.  The file specifies:
1. The number of frequency bands
2. The minimum value of the frequency band in kHz
3. The maximum value of the frequency band in kHz
4. An (ordered) tranmission frequency in this band (kHz)
5. The number of times this transmission frequency was seen in the specified file(s)

The observed frequencies are included alongside the frequency bands to help the user determine if their band width is appropriate for this radar's operating mode.  For the specified DCE file, the output looks like:

```
cat -n dce.freq_band_500 | tail -1
   378	2	12199	12498	12497	46
```

```
head -n 2 dce.freq_band_500
# NBANDS BAND_MIN_KHZ BAND_MAX_KHZ TRANS_FREQ_KHZ NUM_TRANS_FREQ
2	10224	10724	10224	4
```

Before leaving Draft status these things should be addressed in some fashion
------------------------------------------------------------------------------
- [x] I don't know how to include useful `hlpstr.h` and `errstr.h` files.  I tried adding them, but they were overwritten upon compilation.  I have them saved in a save place (include directory) locally, but these files are ignored by Git.
- [x] Is this tool in the right place?
- [x] `grid_utils` and `read_fit_funcs` were taken from other routines.  They should probably be made accessible in another way, since they're useful.
- [x] Make the `get_fit_tfreq_bands` routine accessible without copying files to another directory

ecbland commented:
This looks really useful @aburrell! Answering two of your questions: 

> I don't know how to include useful hlpstr.h and errstr.h files. I tried adding them, but they were overwritten upon compilation.

These files are generated during compilation using the contents of the file `./doc/find_freqband.xml`. This file doesn't exist yet, but once it's there these `.h` files should show up. Have a look at the other routines in `src.bin/tk/tool/` for examples of how to structure this file.

> Is this tool in the right place?

Yes, `src.bin/tk/tool/` seems appropriate.

egthomas commented:
> Is this tool in the right place?

Just another small comment - a `src` subdirectory is not necessary for binaries (in contrast to libraries), so you can place the `makefile` and all of the other `.c` files directly in the `find_freqband.1.0` directory.

aburrell commented on codebase/superdarn/src.bin/tk/tool/find_freqband.1.0/grid_utils.c:
```diff
@@ -0,0 +1,83 @@
+/* make_grid.c
```
```suggestion
/* grid_utils.c
```

aburrell commented:
Thanks @ecbland and @egthomas!  Now for the harder questions on the tick list 😛

aburrell commented on codebase/general/src.lib/time.1.7/include/rtime.h:
```diff
@@ -20,6 +20,8 @@ along with this program. If not, see <https://www.gnu.org/licenses/>.
 
 Modifications:
 
+AGB: Moved string-to-epoch-time routines here from make_grid.2.0 to allow
+     better access across all tools.
```
```suggestion
2021-08-27 Angeline G. Burrell: Moved string-to-epoch-time routines here from make_grid.2.0
     to allow better access across all tools.
```

aburrell commented on codebase/general/src.lib/time.1.7/src/time.c:
```diff
@@ -20,6 +20,9 @@ along with this program. If not, see <https://www.gnu.org/licenses/>.
 
 Modifications:
 
+AGB: Moved string-to-epoch-time routines here from make_grid.2.0 to allow
+     better access across all tools.
```
```suggestion
2021-08-27 Angeline G Burrell: Moved string-to-epoch-time routines here from make_grid.2.0
     to allow better access across all tools.
```

aburrell commented on codebase/superdarn/src.lib/tk/scan.1.7/src/radarscan.c:
```diff
@@ -20,12 +20,15 @@ You should have received a copy of the GNU General Public License
 along with this program. If not, see <https://www.gnu.org/licenses/>.
 
 Modifications:
+AGB - Moved exclude_outofscan routine here to allow access by multiple tools.
```
```suggestion
2021-08-27 Angeline G. Burrell: Moved exclude_outofscan routine here to allow access by multiple tools.
```

ecbland commented:
Sorry to keep you waiting with the PR, @aburrell 

Unfortunately I'm not able to compile RST on this branch. Here's the error message: 
```
================================================================================
Compiling Binary:/rst/codebase/superdarn/src.bin/tk/tool/find_freqband.1.0
================================================================================
make clean
rm -f *.o
rm -f version.h
rm -f errstr.h
rm -f hlpstr.h
rm -f find_freqband
make
make.version /home/emma/gitrepos/rst/codebase/superdarn/src.bin/tk/tool/find_freqband.1.0
make.help
./doc/find_freqband.doc.xml
cc -fPIC -Wall -pedantic -O3 -D_GNU_SOURCE -D_LINUX  -D_SVGLIB_ -I/rst/include/analysis -I/rst/include/base -I/rst/include/general -I/rst/include/superdarn -I../include/ -c find_freqband.c    
cc -fPIC -Wall -pedantic -O3 -D_GNU_SOURCE -D_LINUX  -D_SVGLIB_ -I/rst/include/analysis -I/rst/include/base -I/rst/include/general -I/rst/include/superdarn -I../include/ -c write_tfreq_band.c    
mkdir -p /rst/bin
cc -L/rst/lib -o /rst/bin/find_freqband find_freqband.o write_tfreq_band.o -Wl,-Bstatic \
                           -loldfit.1 -lgtabw.1 -loldgtabw.1 -lgtable.1 -lrpos.1 -lfilter.1 -lcfit.1 -lfit.1 -lrscan.1 -lradar.1 -ldmap.1 -lopt.1 -lrtime.1 -lrcnv.1 -laacgm.1 -ligrf.1 -laacgm_v2.1 -ligrf_v2.1 -lastalg.1 -lfitacf.1 -lchannel.1 -lfreqband.1  -Wl,-Bdynamic \
                           -lm -lz   
/usr/bin/ld: /rst/lib/libfreqband.1.a(get_fit_tfreq_band.o): in function `get_fit_tfreq_bands':
get_fit_tfreq_band.c:(.text+0xa0): undefined reference to `RadarScanMake'
/usr/bin/ld: get_fit_tfreq_band.c:(.text+0xaa): undefined reference to `RadarParmMake'
/usr/bin/ld: get_fit_tfreq_band.c:(.text+0xb6): undefined reference to `FitMake'
/usr/bin/ld: get_fit_tfreq_band.c:(.text+0xc2): undefined reference to `CFitMake'
/usr/bin/ld: get_fit_tfreq_band.c:(.text+0x18b): undefined reference to `FitIndexFload'
/usr/bin/ld: get_fit_tfreq_band.c:(.text+0x1ef): undefined reference to `FitFreadRadarScan'
/usr/bin/ld: get_fit_tfreq_band.c:(.text+0x371): undefined reference to `OldFitReadRadarScan'
/usr/bin/ld: get_fit_tfreq_band.c:(.text+0x3cb): undefined reference to `OldFitClose'
/usr/bin/ld: get_fit_tfreq_band.c:(.text+0x58c): undefined reference to `FitFreadRadarScan'
/usr/bin/ld: get_fit_tfreq_band.c:(.text+0x5c3): undefined reference to `CFitReadRadarScan'
/usr/bin/ld: get_fit_tfreq_band.c:(.text+0x5d4): undefined reference to `exclude_outofscan'
/usr/bin/ld: get_fit_tfreq_band.c:(.text+0x71b): undefined reference to `OldFitSeek'
/usr/bin/ld: get_fit_tfreq_band.c:(.text+0x78d): undefined reference to `FitFread'
/usr/bin/ld: get_fit_tfreq_band.c:(.text+0x7cc): undefined reference to `FitFreadRadarScan'
/usr/bin/ld: get_fit_tfreq_band.c:(.text+0x7ee): undefined reference to `CFitClose'
/usr/bin/ld: get_fit_tfreq_band.c:(.text+0x800): undefined reference to `CFitOpen'
/usr/bin/ld: get_fit_tfreq_band.c:(.text+0x835): undefined reference to `CFitReadRadarScan'
/usr/bin/ld: get_fit_tfreq_band.c:(.text+0x8b0): undefined reference to `OldFitOpen'
/usr/bin/ld: get_fit_tfreq_band.c:(.text+0x8e2): undefined reference to `OldFitReadRadarScan'
/usr/bin/ld: get_fit_tfreq_band.c:(.text+0x93d): undefined reference to `CFitSeek'
/usr/bin/ld: get_fit_tfreq_band.c:(.text+0x985): undefined reference to `CFitReadRadarScan'
/usr/bin/ld: get_fit_tfreq_band.c:(.text+0x9cc): undefined reference to `OldFitReadRadarScan'
/usr/bin/ld: get_fit_tfreq_band.c:(.text+0x9ec): undefined reference to `FitFseek'
/usr/bin/ld: get_fit_tfreq_band.c:(.text+0xa36): undefined reference to `OldFitRead'
/usr/bin/ld: get_fit_tfreq_band.c:(.text+0xaa7): undefined reference to `CFitRead'
collect2: error: ld returned 1 exit status
make: *** [/rst/build/make/makebin.linux:43: find_freqband] Error 1
Compilation Aborted.
```

ecbland commented on codebase/superdarn/src.bin/tk/tool/find_freqband.1.0/find_freqband.c:
```diff
@@ -0,0 +1,244 @@
+ This file is part of the Radar Software Toolkit (RST).
+
+ RST is free software: you can redistribute it and/or modify
+ it under the terms of the GNU Lesser General Public License as published by
```
Did you intend to use the LGPL disclaimer here rather than GPL? See also `channel.c`

aburrell replied:
Ok, I found an example and propagated it.  If it's wrong I'll have another go.

aburrell commented on codebase/superdarn/src.lib/tk/freqband.1.0/src/makefile:
```diff
@@ -0,0 +1,35 @@
+# Makefile for frequency band library
+# ===================================
+# by A.G. Burrell (NRL)
+# Copyright (C) <year>  <name of author>
```
I am not allowed to copyright the files.  Should I just delete the copyright line?

ecbland replied:
Interesting. I'll send you an email in a bit and we can try to find a solution.

aburrell commented:
Ok, I am having a hard time reproducing your compilation error.

```
================================================================================
Compiling Binary:/Users/aburrell/Programs/Git/rst/codebase/superdarn/src.bin/tk/tool/find_freqband.1.0
================================================================================
make clean
rm -f *.o
rm -f version.h
rm -f errstr.h
rm -f hlpstr.h
rm -f find_freqband
make
make.version /Users/aburrell/Programs/Git/rst/codebase/superdarn/src.bin/tk/tool/find_freqband.1.0
make.help
./doc/find_freqband.doc.xml
cc -Wall -pedantic -O3 -D_GNU_SOURCE -D_DARWIN -I/opt/local/include -D_SVGLIB_ -I/Users/aburrell/Programs/Git/rst/include/analysis -I/Users/aburrell/Programs/Git/rst/include/base -I/Users/aburrell/Programs/Git/rst/include/general -I/Users/aburrell/Programs/Git/rst/include/superdarn -I../include/ -c find_freqband.c    
cc -Wall -pedantic -O3 -D_GNU_SOURCE -D_DARWIN -I/opt/local/include -D_SVGLIB_ -I/Users/aburrell/Programs/Git/rst/include/analysis -I/Users/aburrell/Programs/Git/rst/include/base -I/Users/aburrell/Programs/Git/rst/include/general -I/Users/aburrell/Programs/Git/rst/include/superdarn -I../include/ -c write_tfreq_band.c    
mkdir -p /Users/aburrell/Programs/Git/rst/bin
cc -Wl,-rpath,@loader_path/../lib/ -Wl,-rpath,@loader_path/../../lib/ -L/opt/local/lib -L/Users/aburrell/Programs/Git/rst/lib -o /Users/aburrell/Programs/Git/rst/bin/find_freqband find_freqband.o write_tfreq_band.o -loldfit.1 -lgtabw.1 -loldgtabw.1 -lgtable.1 -lrpos.1 -lfilter.1 -lcfit.1 -lfit.1 -lrscan.1 -lradar.1 -ldmap.1 -lopt.1 -lrtime.1 -lrcnv.1 -laacgm.1 -ligrf.1 -laacgm_v2.1 -ligrf_v2.1 -lastalg.1 -lfitacf.1 -lchannel.1 -lfreqband.1  \
                            -lm -lz
```

I removed it all and tried from scratch to be sure.  I also don't see why the command it failed at would throw that error, because I double checked and all the correct libraries are included.  Any ideas what could be up?

egthomas commented:
@aburrell I haven't tried compiling this yet, but often the order of the libraries included in the makefile is important.  You might try moving `-lchannel.1` and `-lfreqband.1` closer to the front of the `LIBS` list in the `find_freqband` makefile.

ecbland commented:
@aburrell After that last commit I'm able to compile this branch :) I will do some proper testing of the routine ASAP (probably tomorrow)

ecbland commented on codebase/superdarn/src.bin/tk/tool/find_freqband.1.0/doc/find_freqband.doc.xml:
```diff
@@ -0,0 +1,91 @@
+</option>
+<option><on><ar>cfitnames</ar></on><od>filenames of the <code>cfit</code> format files.</od>
+</option>
+<synopsis><p>Creates a <code>grid</code> format file for a single radar site from either a <code>fit</code>, <code>fitacf</code> or <code>cfit</code> format file.</p></synopsis>
```
There's quite a few references to the `make_grid` routine and command line options in this file. Should these be removed?

aburrell replied:
Arg... probably not!

ecbland commented on codebase/superdarn/src.bin/tk/tool/find_freqband.1.0/find_freqband.c:
```diff
@@ -0,0 +1,231 @@
+
+   This file is part of the Radar Software Toolkit (RST).
+
+   Disclaimer: This code is licensed under GPL v3.0 please LICENSE to see the
+               full license
```
```suggestion
   Disclaimer: RST is licensed under GPL v3.0. Please visit 
               <https://www.gnu.org/licenses/> to see the full license
```

ecbland commented on codebase/superdarn/src.bin/tk/tool/find_freqband.1.0/write_tfreq_band.c:
```diff
@@ -0,0 +1,56 @@
+
+   This file is part of the Radar Software Toolkit (RST).
+
+   Disclaimer: This code is licensed under GPL v3.0 please LICENSE to see the
+               full license
```
```suggestion
   Disclaimer: RST is licensed under GPL v3.0. Please visit 
               <https://www.gnu.org/licenses/> to see the full license
```

ecbland commented on codebase/superdarn/src.lib/tk/freqband.1.0/src/makefile:
```diff
@@ -0,0 +1,24 @@
+# Author: Angeline G. Burrell, NRL (2021)
+# This is a U.S. government work and not under copyright protection in the U.S.
+#
+# Disclaimer: This code is licensed under GPL v3.0 please LICENSE to 
+#             see the full license
```
```suggestion
# Disclaimer: RST is licensed under GPL v3.0. Please visit 
#             <https://www.gnu.org/licenses/> to see the full license
```

ecbland reviewed (commented):
@aburrell I've tested the code and it works as described. Here's an example for LYR: 

```
# NBANDS BAND_MIN_KHZ BAND_MAX_KHZ TRANS_FREQ_KHZ NUM_TRANS_FREQ
2	 9802	10102	 9802	220
2	 9802	10102	 9803	107
2	 9802	10102	 9804	58
2	 9802	10102	 9805	60
2	 9802	10102	 9806	36
2	 9802	10102	 9807	36
2	 9802	10102	 9808	30
2	 9802	10102	 9809	14
2	 9802	10102	 9810	15
2	 9802	10102	 9811	20
2	 9802	10102	 9812	8
2	 9802	10102	 9813	10
2	 9802	10102	 9814	13
2	 9802	10102	 9815	7
2	 9802	10102	 9816	3
2	 9802	10102	 9817	3
2	 9802	10102	 9881	1
2	 9802	10102	 9882	5
2	 9802	10102	 9883	8
2	 9802	10102	 9884	7
2	 9802	10102	 9885	17
2	 9802	10102	 9886	28
2	 9802	10102	 9887	35
2	 9802	10102	 9888	49
2	 9802	10102	 9889	59
2	 9802	10102	 9890	80
2	 9802	10102	 9891	112
2	 9802	10102	 9892	120
2	 9802	10102	 9893	141
2	 9802	10102	 9894	140
2	 9802	10102	 9895	173
2	 9802	10102	 9896	305
2	11602	11697	11602	194
2	11602	11697	11603	87
2	11602	11697	11604	69
2	11602	11697	11605	82
2	11602	11697	11606	91
2	11602	11697	11607	65
2	11602	11697	11608	68
2	11602	11697	11609	29
2	11602	11697	11610	24
2	11602	11697	11611	18
2	11602	11697	11612	18
2	11602	11697	11613	4
2	11602	11697	11614	7
2	11602	11697	11615	2
2	11602	11697	11616	6
2	11602	11697	11617	2
2	11602	11697	11683	3
2	11602	11697	11684	2
2	11602	11697	11685	6
2	11602	11697	11686	10
2	11602	11697	11687	15
2	11602	11697	11688	30
2	11602	11697	11689	37
2	11602	11697	11690	39
2	11602	11697	11691	66
2	11602	11697	11692	86
2	11602	11697	11693	123
2	11602	11697	11694	176
2	11602	11697	11695	149
2	11602	11697	11696	412
```
For anyone who is interested: Notice how the frequencies with highest occurrence (`NUM_TRANS_FREQ`) are at the edges of the bands? This happens because the narrow IF filters suppress the noise at the edges of each band, and the clear frequency search does not account for this when choosing a quiet frequency. This is known behaviour for LYR (and elsewhere), so the above result is the expected one :smile: 

Just a few quick comments on the documentation and license notices, and then I will merge this.

aburrell commented:
Ok @ecbland I searched for 'grid' and 'grd' in the 'find_frequency' documentation and think I removed all the places the old 'make_grid' information was lingering.  I also updated all of the new licenses to the format you corrected.
